Bug: Periodic Failure of Consumer and Producer tests.

CI Output:

```
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"OQ==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"MA==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"Mg==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"MQ==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"Mw==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"NA==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=NotifyClose ID=default INSTANCE=3 OP_ATTEMPT=-1 TRANSPORT=CHANNEL zARGS="marshaling error: json: unsupported type: chan *amqp.Error" zRESULTS="marshaling error: json: unsupported type: chan *amqp.Error"
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"Ng==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"NQ==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"OA==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
7:41PM DBG  :METHOD_CALL=Publish ID=default INSTANCE=3 OP_ATTEMPT=0 TRANSPORT=CHANNEL zARGS={"Exchange":"","Immediate":false,"Key":"test_queue_producer_publish","Mandatory":false,"Msg":{"AppId":"","Body":"Nw==","ContentEncoding":"","ContentType":"","CorrelationId":"","DeliveryMode":0,"Expiration":"","Headers":null,"MessageId":"","Priority":0,"ReplyTo":"","Timestamp":"0001-01-01T00:00:00Z","Type":"","UserId":""}}
panic: confirmation delivery tag 3 does not line up with order delivery tag 1

goroutine 52 [running]:
github.com/peake100/rogerRabbit-go/roger/amqpproducer.(*Producer).runConfirmations(0xc000296a00)
	/home/vsts/work/1/s/roger/amqpproducer/producer.go:100 +0x51c
github.com/peake100/rogerRabbit-go/roger/amqpproducer.(*Producer).Run.func3(0xc0003bb2e0, 0xc000296a00)
	/home/vsts/work/1/s/roger/amqpproducer/producer.go:295 +0x88
created by github.com/peake100/rogerRabbit-go/roger/amqpproducer.(*Producer).Run
	/home/vsts/work/1/s/roger/amqpproducer/producer.go:293 +0x39b
FAIL	github.com/peake100/rogerRabbit-go/roger/amqpproducer	0.063s
```

--- a/amqp/channelEventRelays.go
+++ b/amqp/channelEventRelays.go
@@ -46,17 +46,26 @@ func shutdownRelay(relay eventRelay, relaySync relaySync) {
 func (channel *Channel) eventRelaySetupAndLaunch(relay eventRelay) {
 	// Launch the runner
 	thisSync := newRelaySync(channel.ctx)
-	go channel.runEventRelay(relay, thisSync)
+
+	setupComplete := make(chan struct{})
+	// Launch the relay.
+	go channel.runEventRelay(relay, thisSync, setupComplete)
+	// Wait for the signal that our setup is complete.
+	<-setupComplete
 }
 
 // runEventRelay should be launched as goroutine to run an event relay after it's
 // initial setup.
-func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync) {
+func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync, setupComplete chan struct{}) {
 	// Shutdown our relay on exit.
 	defer shutdownRelay(relay, relaySync)
 
 	firstLegComplete := make(chan struct{})
-	channel.eventRelayInitialSetup(relay, relaySync, firstLegComplete)
+	func() {
+		// Close the setupComplete so we can return to the caller.
+		defer close(setupComplete)
+		channel.eventRelayInitialSetup(relay, relaySync, firstLegComplete)
+	}()
 
 	// Wait for ou first leg to complete, then fall into a rhythm with the transport
 	// manager

--- a/amqp/channelEventRelays.go
+++ b/amqp/channelEventRelays.go
@@ -56,7 +56,7 @@ func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync) {
 	defer shutdownRelay(relay, relaySync)
 
 	firstLegComplete := make(chan struct{})
-	channel.eventRelaySetup(relay, relaySync, firstLegComplete)
+	channel.eventRelayInitialSetup(relay, relaySync, firstLegComplete)
 
 	// Wait for ou first leg to complete, then fall into a rhythm with the transport
 	// manager
@@ -73,7 +73,11 @@ func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync) {
 	}
 }
 
-func (channel *Channel) eventRelaySetup(
+// eventRelayInitialSetup does the initial setup of an event relay. We need to make sure
+// that setup is complete before we return to the user, otherwise they may start taking
+// actions that SHOULD trigger events before the event listener has been registered with
+// the underlying streadway/amqp.Channel.
+func (channel *Channel) eventRelayInitialSetup(
 	relay eventRelay, relaySync relaySync, firstLegComplete chan struct{},
 ) {
 	// Signal this leg in an op so we can make sure we grab the right channel.

--- a/amqp/channelEventRelays.go
+++ b/amqp/channelEventRelays.go
@@ -44,12 +44,17 @@ func shutdownRelay(relay eventRelay, relaySync relaySync) {
 
 // eventRelaySetupAndLaunch sets up a new relay and launches a goroutine to run it
 func (channel *Channel) eventRelaySetupAndLaunch(relay eventRelay) {
-	// Launch the runner
-	thisSync := newRelaySync(channel.ctx)
-
+	// Create a signal chan to tell us when the initial setup has been completed. This
+	// means it is safe to return to the user as our listener channel has been
+	// registered with the underlying channel. If we were to return immediately, the
+	// user might start taking actions that SHOULD generate events before the even
+	// channel is correctly registered, causing those events to go "missing".
 	setupComplete := make(chan struct{})
+
 	// Launch the relay.
+	thisSync := newRelaySync(channel.ctx)
 	go channel.runEventRelay(relay, thisSync, setupComplete)
+
 	// Wait for the signal that our setup is complete.
 	<-setupComplete
 }
@@ -61,11 +66,7 @@ func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync, set
 	defer shutdownRelay(relay, relaySync)
 
 	firstLegComplete := make(chan struct{})
-	func() {
-		// Close the setupComplete so we can return to the caller.
-		defer close(setupComplete)
-		channel.eventRelayInitialSetup(relay, relaySync, firstLegComplete)
-	}()
+	channel.eventRelayInitialSetup(relay, relaySync, setupComplete, firstLegComplete)
 
 	// Wait for ou first leg to complete, then fall into a rhythm with the transport
 	// manager
@@ -87,7 +88,10 @@ func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync, set
 // actions that SHOULD trigger events before the event listener has been registered with
 // the underlying streadway/amqp.Channel.
 func (channel *Channel) eventRelayInitialSetup(
-	relay eventRelay, relaySync relaySync, firstLegComplete chan struct{},
+	relay eventRelay,
+	relaySync relaySync,
+	setupComplete chan struct{},
+	firstLegComplete chan struct{},
 ) {
 	// Signal this leg in an op so we can make sure we grab the right channel.
 	_ = channel.transportManager.retryOperationOnClosed(
@@ -95,8 +99,6 @@ func (channel *Channel) eventRelayInitialSetup(
 		func(ctx context.Context) error {
 			// Register the relay with the channel.
 			channel.relaySync.AddRelay(relaySync.shared)
-
-			setupComplete := make(chan struct{})
 
 			// Run the fist leg with the current channel. We need to launch it in a
 			// routine so we can signal leg complete (the manager needs to grab a write

--- a/amqp/channelEventRelays.go
+++ b/amqp/channelEventRelays.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"github.com/peake100/rogerRabbit-go/amqp/amqpmiddleware"
 	streadway "github.com/streadway/amqp"
-	"sync"
 )
 
 // createEventMetadata creates the amqpmiddleware.EventMetadata for an event
@@ -17,7 +16,7 @@ func createEventMetadata(legNum int, eventNum int64) amqpmiddleware.EventMetadat
 
 // eventRelay is a common interface for relaying events from the underlying channels to
 // the client without interruption. The boilerplate of handling all the synchronization
-// locks will be handled for any relay passed to Channel.setupAndLaunchEventRelay()
+// locks will be handled for any relay passed to Channel.eventRelaySetupAndLaunch()
 type eventRelay interface {
 	// SetupForRelayLeg runs the setup for a new relay leg.
 	SetupForRelayLeg(newChannel *streadway.Channel) error
@@ -41,6 +40,83 @@ func shutdownRelay(relay eventRelay, relaySync relaySync) {
 
 	// Invoke the shutdown method of the relay.
 	_ = relay.Shutdown()
+}
+
+// eventRelaySetupAndLaunch sets up a new relay and launches a goroutine to run it
+func (channel *Channel) eventRelaySetupAndLaunch(relay eventRelay) {
+	// Launch the runner
+	thisSync := newRelaySync(channel.ctx)
+	go channel.runEventRelay(relay, thisSync)
+}
+
+// runEventRelay should be launched as goroutine to run an event relay after it's
+// initial setup.
+func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync) {
+	// Shutdown our relay on exit.
+	defer shutdownRelay(relay, relaySync)
+
+	firstLegComplete := make(chan struct{})
+	channel.eventRelaySetup(relay, relaySync, firstLegComplete)
+
+	// Wait for ou first leg to complete, then fall into a rhythm with the transport
+	// manager
+	<-firstLegComplete
+	relayLeg := 1
+
+	// Start running each leg.
+	for {
+		channel.runEventRelayCycle(relay, relaySync, relayLeg)
+		if relaySync.IsDone() {
+			return
+		}
+		relayLeg++
+	}
+}
+
+func (channel *Channel) eventRelaySetup(
+	relay eventRelay, relaySync relaySync, firstLegComplete chan struct{},
+) {
+	// Signal this leg in an op so we can make sure we grab the right channel.
+	_ = channel.transportManager.retryOperationOnClosed(
+		channel.ctx,
+		func(ctx context.Context) error {
+			// Register the relay with the channel.
+			channel.relaySync.AddRelay(relaySync.shared)
+
+			setupComplete := make(chan struct{})
+
+			// Run the fist leg with the current channel. We need to launch it in a
+			// routine so we can signal leg complete (the manager needs to grab a write
+			// lock to the transport before it checks the relays)
+			go func(currentChannel *streadway.Channel) {
+				defer close(firstLegComplete)
+				defer relaySync.SignalLegComplete()
+
+				// Run the relay setup then signal that initial setup is complete.
+				var err error
+				func() {
+					defer close(setupComplete)
+					err = relay.SetupForRelayLeg(currentChannel)
+				}()
+				if err != nil {
+					return
+				}
+
+				// Run the first relay leg.
+				if done := relay.RunRelayLeg(0); done {
+					relaySync.SetDone()
+				}
+			}(channel.underlyingChannel)
+
+			// Wait for the relay setup to complete before we return or release the read
+			// lock to the user. Otherwise the user may think we are receiving events
+			// and do something that creates them before we've actually set that up.
+			<-setupComplete
+
+			return nil
+		},
+		true,
+	)
 }
 
 // runEventRelayCycle runs a single, full cycle of setting up and running a relay leg.
@@ -68,63 +144,4 @@ func (channel *Channel) runEventRelayCycle(
 	if done := relay.RunRelayLeg(legNum); done {
 		relaySync.SetDone()
 	}
-}
-
-// runEventRelay should be launched as goroutine to run an event relay after it's
-// initial setup.
-func (channel *Channel) runEventRelay(relay eventRelay, relaySync relaySync) {
-	// Shutdown our relay on exit.
-	defer shutdownRelay(relay, relaySync)
-
-	firstLegComplete := new(sync.WaitGroup)
-	firstLegComplete.Add(1)
-
-	// Signal this leg in an op so we can make sure we grab the right channel.
-	_ = channel.transportManager.retryOperationOnClosed(
-		channel.ctx,
-		func(ctx context.Context) error {
-			// Register the relay with the channel.
-			channel.relaySync.AddRelay(relaySync.shared)
-
-			// Run the fist leg with the current channel. We need to launch it in a
-			// routine so we can signal leg complete (the manager needs to grab a write
-			// lock to the transport before it checks the relays)
-			go func(currentChannel *streadway.Channel) {
-				defer firstLegComplete.Done()
-				defer relaySync.SignalLegComplete()
-
-				err := relay.SetupForRelayLeg(currentChannel)
-				if err != nil {
-					return
-				}
-
-				if done := relay.RunRelayLeg(0); done {
-					relaySync.SetDone()
-				}
-			}(channel.underlyingChannel)
-			return nil
-		},
-		true,
-	)
-
-	// Wait for ou first leg to complete, then fall into a rhythm with the transport
-	// manager
-	firstLegComplete.Wait()
-	relayLeg := 1
-
-	// Start running each leg.
-	for {
-		channel.runEventRelayCycle(relay, relaySync, relayLeg)
-		if relaySync.IsDone() {
-			return
-		}
-		relayLeg++
-	}
-}
-
-// setupAndLaunchEventRelay sets up a new relay and launches a goroutine to run it
-func (channel *Channel) setupAndLaunchEventRelay(relay eventRelay) {
-	// Launch the runner
-	thisSync := newRelaySync(channel.ctx)
-	go channel.runEventRelay(relay, thisSync)
 }

--- a/amqp/channelHandlersBuilder.go
+++ b/amqp/channelHandlersBuilder.go
@@ -666,7 +666,6 @@ func (builder channelHandlerBuilder) createNotifyReturn() amqpmiddleware.Handler
 
 	handler := func(ctx context.Context, args amqpmiddleware.ArgsNotifyReturn) amqpmiddleware.ResultsNotifyReturn {
 		relay := newNotifyReturnRelay(args.Returns, eventMiddlewares)
-
 		channel.eventRelaySetupAndLaunch(relay)
 		return amqpmiddleware.ResultsNotifyReturn{Returns: args.Returns}
 	}

--- a/amqp/channelHandlersBuilder.go
+++ b/amqp/channelHandlersBuilder.go
@@ -458,7 +458,7 @@ func (builder channelHandlerBuilder) createConsume() amqpmiddleware.HandlerConsu
 		relay := newConsumeRelay(callArgs, channel, eventMiddleware)
 
 		// Pass it to our relay handler.
-		channel.setupAndLaunchEventRelay(relay)
+		channel.eventRelaySetupAndLaunch(relay)
 
 		results.DeliveryChan = callArgs.callerDeliveryChan
 		// If no error, pass the channel back to the caller
@@ -537,7 +537,7 @@ func (builder channelHandlerBuilder) createNotifyPublish() amqpmiddleware.Handle
 	) amqpmiddleware.ResultsNotifyPublish {
 		relay := newNotifyPublishRelay(args.Confirm, eventMiddleware)
 
-		channel.setupAndLaunchEventRelay(relay)
+		channel.eventRelaySetupAndLaunch(relay)
 
 		return amqpmiddleware.ResultsNotifyPublish{Confirm: args.Confirm}
 	}
@@ -667,7 +667,7 @@ func (builder channelHandlerBuilder) createNotifyReturn() amqpmiddleware.Handler
 	handler := func(ctx context.Context, args amqpmiddleware.ArgsNotifyReturn) amqpmiddleware.ResultsNotifyReturn {
 		relay := newNotifyReturnRelay(args.Returns, eventMiddlewares)
 
-		channel.setupAndLaunchEventRelay(relay)
+		channel.eventRelaySetupAndLaunch(relay)
 		return amqpmiddleware.ResultsNotifyReturn{Returns: args.Returns}
 	}
 
@@ -683,7 +683,7 @@ func (builder channelHandlerBuilder) createNotifyCancel() amqpmiddleware.Handler
 	handler := func(ctx context.Context, args amqpmiddleware.ArgsNotifyCancel) amqpmiddleware.ResultsNotifyCancel {
 		relay := newNotifyCancelRelay(args.Cancellations, eventMiddlewares)
 
-		channel.setupAndLaunchEventRelay(relay)
+		channel.eventRelaySetupAndLaunch(relay)
 		return amqpmiddleware.ResultsNotifyCancel{Cancellations: args.Cancellations}
 	}
 
@@ -705,7 +705,7 @@ func (builder channelHandlerBuilder) createNotifyFlow() amqpmiddleware.HandlerNo
 		)
 
 		// Setup and launch the relay.
-		channel.setupAndLaunchEventRelay(relay)
+		channel.eventRelaySetupAndLaunch(relay)
 		return amqpmiddleware.ResultsNotifyFlow{FlowNotifications: args.FlowNotifications}
 	}
 

--- a/amqp/channel_test.go
+++ b/amqp/channel_test.go
@@ -934,8 +934,6 @@ func (suite *ChannelMethodsSuite) Test0160_NotifyConfirm() {
 
 	suite.ChannelPublish().NotifyConfirm(ackEvents, nackEvents)
 
-	confirmations := new(sync.WaitGroup)
-
 	go func() {
 		for i := 0; i < publishCount; i++ {
 			err := suite.ChannelPublish().Publish(
@@ -950,8 +948,6 @@ func (suite *ChannelMethodsSuite) Test0160_NotifyConfirm() {
 			if !suite.NoErrorf(err, "publish %v", i) {
 				suite.T().FailNow()
 			}
-			// Add 1 to the confirmation WaitGroup
-			confirmations.Add(1)
 		}
 	}()
 
@@ -983,7 +979,7 @@ func (suite *ChannelMethodsSuite) Test0160_NotifyConfirm() {
 			suite.T().FailNow()
 		}
 
-		if receivedCount >= 10 {
+		if receivedCount >= publishCount {
 			break
 		}
 	}

--- a/amqp/transportManager.go
+++ b/amqp/transportManager.go
@@ -333,7 +333,7 @@ func (manager *transportManager) retryOperationOnClosed(
 		//
 		// We'll give one immediate retry, but after that start increasing how long
 		// we need to wait before re-attempting.
-		waitDur := 5 * time.Millisecond * time.Duration(attempt - 1)
+		waitDur := 5 * time.Millisecond * time.Duration(attempt-1)
 		if waitDur > maxWait {
 			waitDur = maxWait
 		}
@@ -509,11 +509,11 @@ func (manager *transportManager) setup(
 	transport reconnects,
 	middleware transportManagerMiddleware,
 ) {
-	ctx, cancelFunc := context.WithCancel(ctx)
+	newCtx, cancelFunc := context.WithCancel(ctx)
 
 	reconnectCount := uint64(0)
 
-	manager.ctx = ctx
+	manager.ctx = newCtx
 	manager.cancelFunc = cancelFunc
 	manager.transport = transport
 	manager.transportLock = new(sync.RWMutex)

--- a/amqp/transportManager.go
+++ b/amqp/transportManager.go
@@ -299,6 +299,8 @@ func (manager *transportManager) retryOperationOnClosedSingle(
 	return err
 }
 
+const maxWait = time.Second * 10
+
 // revive:enable:context-as-argument
 
 // retryOperationOnClosed repeats operation / method call until a non-closed error is
@@ -323,6 +325,17 @@ func (manager *transportManager) retryOperationOnClosed(
 
 		// Otherwise retry the operation once the connection has been established.
 		attempt++
+
+		// We don't want to saturate the connection with retries if we are having
+		// a hard time reconnecting.
+		//
+		// We'll give one immediate retry, but after that start increasing how long
+		// we need to wait before re-attempting.
+		waitDur := 5 * time.Millisecond * time.Duration(attempt - 1)
+		if waitDur > time.Second * 30 {
+			waitDur = maxWait
+		}
+		time.Sleep(waitDur)
 	}
 }
 

--- a/amqp/transportManager.go
+++ b/amqp/transportManager.go
@@ -299,6 +299,8 @@ func (manager *transportManager) retryOperationOnClosedSingle(
 	return err
 }
 
+// maxWait is the max amount of time a method should wait before trying again after
+// multiple failures in a row.
 const maxWait = time.Second * 5
 
 // revive:enable:context-as-argument

--- a/amqp/transportManager.go
+++ b/amqp/transportManager.go
@@ -299,7 +299,7 @@ func (manager *transportManager) retryOperationOnClosedSingle(
 	return err
 }
 
-const maxWait = time.Second * 10
+const maxWait = time.Second * 5
 
 // revive:enable:context-as-argument
 
@@ -332,7 +332,7 @@ func (manager *transportManager) retryOperationOnClosed(
 		// We'll give one immediate retry, but after that start increasing how long
 		// we need to wait before re-attempting.
 		waitDur := 5 * time.Millisecond * time.Duration(attempt - 1)
-		if waitDur > time.Second * 30 {
+		if waitDur > maxWait {
 			waitDur = maxWait
 		}
 		time.Sleep(waitDur)

--- a/amqp/transportManagerReconnect.go
+++ b/amqp/transportManagerReconnect.go
@@ -48,11 +48,14 @@ func (manager *transportManager) reconnectRedial(
 
 // reconnectListenForClose listens for a close event from the underlying livesOnce, and
 // starts the reconnection process.
-func (manager *transportManager) reconnectListenForClose(
-	closeChan <-chan *streadway.Error,
-) {
+func (manager *transportManager) reconnectListenForClose(closeChan <-chan *streadway.Error) {
 	// Wait for the current connection to close
 	disconnectEvent := <-closeChan
+
+	// Lock access to the connection and don't unlock until we have reconnected.
+	manager.transportLock.Lock()
+	defer manager.transportLock.Unlock()
+
 	// Send a disconnect event to all interested subscribers.
 	manager.sendDisconnectNotifications(disconnectEvent)
 
@@ -69,9 +72,12 @@ func (manager *transportManager) reconnectListenForClose(
 // reconnect establishes a new underlying connection and sets up a listener for it's
 // closure.
 func (manager *transportManager) reconnect(ctx context.Context, retry bool) error {
-	// Lock access to the connection and don't unlock until we have reconnected.
-	manager.transportLock.Lock()
-	defer manager.transportLock.Unlock()
+	// This may be called directly by Dial methods. It's okay NOT to use the lock here
+	// since the caller won't be handed back the Connection or Channel until the initial
+	// one is established.
+	//
+	// Once the first connection is established, reconnectListenForClose will grab
+	// the lock immediately on a disconnect.
 
 	// Redial the broker until we reconnectMiddleware
 	err := manager.reconnectRedial(ctx, retry)
@@ -79,12 +85,12 @@ func (manager *transportManager) reconnect(ctx context.Context, retry bool) erro
 		return err
 	}
 
-	// Broadcast that we have made a successful reconnection to any one-time listeners.
-	manager.reconnectCond.Broadcast()
-
 	// Register a notification channelConsume for the new connection's closure.
 	closeChan := make(chan *streadway.Error, 1)
 	manager.transport.underlyingTransport().NotifyClose(closeChan)
+
+	// Broadcast that we have made a successful reconnection to any one-time listeners.
+	manager.reconnectCond.Broadcast()
 
 	// Launch a goroutine to reconnectMiddleware on connection closure.
 	go manager.reconnectListenForClose(closeChan)

--- a/amqptest/testSuite.go
+++ b/amqptest/testSuite.go
@@ -360,9 +360,10 @@ func (amqpSuite *AmqpSuite) publishMessagesSend(
 ) {
 	assert := assert.New(t)
 
+	channel := amqpSuite.ChannelPublish()
 	for i := 0; i < count; i++ {
 		msg := strconv.Itoa(i)
-		err := amqpSuite.ChannelPublish().Publish(
+		err := channel.Publish(
 			exchange,
 			key,
 			true,
@@ -397,6 +398,7 @@ func (amqpSuite *AmqpSuite) publishMessagesConfirm(
 				allConfirmed <- fmt.Errorf("message %v nacked", confirmCount)
 				return
 			}
+			confirmCount++
 		case <-returns:
 			t.Error("message returned by broker")
 			allConfirmed <- errors.New("message returned by broker")
@@ -404,7 +406,6 @@ func (amqpSuite *AmqpSuite) publishMessagesConfirm(
 		case <-ctx.Done():
 			return
 		}
-		confirmCount++
 		if confirmCount == count {
 			return
 		}

--- a/roger/amqpproducer/producer_test.go
+++ b/roger/amqpproducer/producer_test.go
@@ -122,6 +122,15 @@ func (suite *ProducerSuite) TestProducerPublish() {
 		suite.T().FailNow()
 	}
 
+	queueInfo, err := suite.ChannelPublish().QueueInspect(queueName)
+	if !suite.NoError(err, "inspect queue") {
+		suite.T().FailNow()
+	}
+
+	if !suite.Equal(publishCount, queueInfo.Messages, "queued message count") {
+		suite.T().FailNow()
+	}
+
 	messages := make([]bool, publishCount)
 	for i := 0; i < publishCount; i++ {
 		msg, ok, err := suite.ChannelPublish().Get(queueName, true)
@@ -187,13 +196,11 @@ func (suite *ProducerSuite) TestProducerQueuePublication() {
 	}()
 
 	for i, thisPublication := range publications {
-		func() {
+		go func(index int, p *amqpproducer.Publication) {
 			defer publishWork.Done()
-			err := thisPublication.WaitOnConfirmation()
-			if !suite.NoError(err, "wait on publication %v", i) {
-				suite.T().FailNow()
-			}
-		}()
+			err := p.WaitOnConfirmation()
+			suite.NoError(err, "wait on publication %v", index)
+		}(i, thisPublication)
 	}
 
 	timer := time.NewTimer(15 * time.Second)
@@ -202,6 +209,15 @@ func (suite *ProducerSuite) TestProducerQueuePublication() {
 	case <-publishComplete:
 	case <-timer.C:
 		suite.T().Error("message publication timeout")
+		suite.T().FailNow()
+	}
+
+	queueInfo, err := suite.ChannelPublish().QueueInspect(queueName)
+	if !suite.NoError(err, "inspect queue") {
+		suite.T().FailNow()
+	}
+
+	if !suite.Equal(publishCount, queueInfo.Messages, "queued message count") {
 		suite.T().FailNow()
 	}
 

